### PR TITLE
E2e registration testing BPoR

### PR DESCRIPTION
**Pull request by ThomasJackSmolen**
Added NamePage.ts POM. 
Added NamePage to CustomFixtures.ts file test object
Added e2e user signup journey test file

### CODE REVIEW

**Comment by paddyduff31** [comment]
@ThomasJackSmolen close this PR and setup a new one pointing to the original env

File: e2e/pages/StageOne/NamePage.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    this.lastNameError = page.getByText("Enter your last name");
+    this.continueButton = page.getByRole("button", { name: "Continue" });
+  }
+
+  // methods
+  async goTo() {
+    await this.page.goto("Participants/Register/Questions");
+  }
+
+  async componentsVisible() {
```
Comment (paddyduff31) [inline e2e/pages/StageOne/NamePage.ts:39]: update this to be an assertion so `assertComponentsVisible`

File: e2e/pages/StageOne/NamePage.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    await expect(this.firstNameLabel).toBeVisible();
+    await expect(this.firstNameInput).toBeVisible();
+    await expect(this.firstNameError).toBeHidden();
+    await expect(this.lastNameLabel).toBeVisible();
+    await expect(this.lastNameInput).toBeVisible();
+    await expect(this.lastNameError).toBeHidden();
+    await expect(this.continueButton).toBeVisible();
+  }
+
+  async correctContent() {
```
Comment (paddyduff31) [inline e2e/pages/StageOne/NamePage.ts:52]: assert

File: e2e/pages/StageOne/NamePage.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      await expect(this.lastNameError).toBeHidden();
+    }
+  }
+
+  async clickContinue() {
+    await this.continueButton.click();
+    await this.validateInputs();
+    // Go to next page if no errors found
+  }
+}
```
Comment (paddyduff31) [inline e2e/pages/StageOne/NamePage.ts:100]: This POM needs a method to clear the input field

File: e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { test } from "../../../../hooks/CustomFixtures";
+
```
Comment (paddyduff31) [inline e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts:2]: This needs built out a little, there is no check if the errors come up, no checks if the first name and last name are in the input field as expected, etc

File: e2e/pages/StageOne/NamePage.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    await this.firstNameInput.fill(firstName);
+  }
+
+  async fillLastName(lastName: string) {
+    await this.lastNameInput.click();
+    await expect(this.lastNameInput).toBeFocused();
+    await this.lastNameInput.fill(lastName);
+  }
+
+  async validateInputs() {
```
Comment (paddyduff31) [inline e2e/pages/StageOne/NamePage.ts:77]: This method is a bit of a mess/ hard to follow, give this another pass and refactor

File: e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+test.describe("Name page journey", () => {
+  test("Name page renders correctly", async ({ namePage }) => {
+    await test.step("I have navigated to the name page", async () => {
```
Comment (paddyduff31) [inline e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts:5]: Follow the wording in other tests ie `Given I have` etc, just good convention

File: e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts (hunk 4 above)
Comment (paddyduff31) [reply e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts:2]: try and cover every possible user action

**Review by paddyduff31 — CHANGES_REQUESTED** [review]
really good first pass, just a bit more work needed

File: e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    await test.step("I will then go back to check that the errors are hidden", async () => {
+      await namePage.clickBack();
+    });
+  });
+});
+
+// Unsuccessful DoB page Journeys
+test.describe("Unsuccessful DoB page journeys", () => {
+  test("Empty day and month fields", async ({ namePage, dateOfBirthPage }) => {
+    await test.step("I will correctly navigate through the name page", async () => {
```
Comment (chrismcneill89) [inline e2e/tests/features/journeys/registration/nhsSignupJourneyTest.e2e.ts:184]: This step is a little too bulky. 
Try and keep things as segmented and separate as possible around 1-3 methods/actions per step.
For this I'd suggest: 
Given I am on the Name Page (.goTo & assertFormVisible)
And I complete the Form with First and Last Name (Fill & Click)

**Review by chrismcneill89 — COMMENTED** [review]
Some comments left for small improvements, nothing major mostly best practice type stuff

File: e2e/pages/StageOne/ConsentPage.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      "No, I do not consent and wish to cancel this registration"
+    );
+    this.backButton = page.getByTitle("Return to previous page");
+    // errors
+    this.checkboxErrorMessage = page.locator("span#checkbox--error-message");
+  }
+
+  // LOAD METHODS
+  async waitForPageLoad() {
+    await this.page.waitForSelector(
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/ConsentPage.ts:81]: Don't use the page.waitForSelector method - https://playwright.dev/docs/api/class-page#page-wait-for-selector
It's been replaced with the much more reliable locator.waitFor method - https://playwright.dev/docs/api/class-locator#locator-wait-for
You have already declared the same locator above, this.consentPageHeading, so use that

Or you could use your assertComponentsVisible method and pass it in
Or even just expect( this.consentPageHeading).toBeVisible()

File: e2e/pages/StageOne/EmailPage.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
+import { Locator, Page, expect } from "@playwright/test";
```
[the remark this answers is not in the document]
Comment (chrismcneill89) [reply e2e/pages/StageOne/EmailPage.ts:1]: Good stuff, though as mentioned on the utils comments just make them generic. 
And then 
// --- ERRORS LOGIC --- //
  async assertError(message: string) {
    await assertErrorUtil(this.errorMessage, message);
  }
  Becomes 
  // --- ERRORS LOGIC --- //
  async assertError(message: string) {
    await assertComponentWithText(this.errorMessage, message);
  }
  
  And
  async assertErrorsHidden() {
    await assertErrorsHidden(this.errorMessage);
  }
  Becomes
  async assertErrorsHidden() {
    await assertComponentsHidden([this.errorMessage]);
  }

File: e2e/pages/StageOne/ConsentPage.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
+import { Locator, Page } from "@playwright/test";
+import { assertComponentsVisible } from "../../utils/visibilityUtils";
+import { assertErrorUtil, assertErrorsHidden } from "../../utils/errorUtils";
+
+export default class ConsentPage {
+  readonly page: Page;
+  // text
+  readonly progressDisplay: Locator;
+  readonly consentPageHeading: Locator;
+  readonly consentParagraph1: Locator;
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/ConsentPage.ts:10]: Try and avoid using numbers in your variable/locator names if possible.
If you can't be more specific then use 'consentParagraphFirst', bePartOfResearchLinkFirst etc instead

File: e2e/pages/StageOne/EmailVerificationPage.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
+import { Locator, Page } from "@playwright/test";
+import { assertComponentsVisible } from "../../utils/visibilityUtils";
+import { assertErrorUtil, assertErrorsHidden } from "../../utils/errorUtils";
+
+export default class EmailVerificationPage {
+  readonly page: Page;
+  // text
+  readonly checkEmailHeading: Locator;
+  readonly emailSentMessage: Locator;
+  readonly text1: Locator;
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/EmailVerificationPage.ts:10]: Same as comment on ConsentPage, if can't be more specific then use 'first', 'second' etc instead of 1,2

File: e2e/pages/StageOne/EmailVerificationPage.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+      "We've resent the email. If you need help, please contact bepartofresearch@nihr.ac.uk"
+    );
+    // button
+    this.resendVerificationEmailButton = page.getByRole("button", {
+      name: "Resend verification email",
+    });
+  }
+
+  // LOAD METHODS
+  async waitForPageLoad() {
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/EmailVerificationPage.ts:46]: Same as ConsentPage, don't use the page.waitForSelector method, use one of the better alternatives

File: e2e/pages/StageOne/NamePage.ts [hunk 13]
```diff
@@ -76,51 +71,8 @@ export default class NamePage {
...
-
-  async assertLastNameError() {
-    await expect(this.firstNameError).toBeHidden();
-    await expect(this.lastNameError).toBeVisible();
-    await expect(this.lastNameError).toHaveText(/Enter your last name/);
-  }
-
-  async assertHideErrorMessages() {
+  // Check no error messages visible before continue button pressed
+  async assertErrorsHidden() {
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/NamePage.ts:75]: Nothing wrong with this, but as mentioned in the utils comments,
If you change the Hidden util method to be generic, and allow it to accept an array like your assertComponentsVisible method.
Then this can become:
  async assertErrorsHidden() {
      await assertComponentsHidden([this.firstNameError, this.lastNameError])
  }

File: e2e/pages/StageOne/PasswordPolicyPage.ts [hunk 14]
```diff
@@ -1,16 +1,52 @@
...
+import { assertErrorUtil, assertErrorsHidden } from "../../utils/errorUtils";
 
 export default class PasswordPolicyPage {
   readonly page: Page;
   // text
   readonly passwordPolicyHeading: Locator;
+  readonly backButton: Locator;
+  readonly continueButton: Locator;
+  readonly progressDisplay: Locator;
+  readonly passwordText1: Locator;
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/PasswordPolicyPage.ts:12]: Same as before, try to not use numbers in variable names, use 'first' etc if needed

File: e2e/pages/StageOne/PasswordPolicyPage.ts [hunk 15]
```diff
@@ -22,4 +58,53 @@ export default class PasswordPolicyPage {
...
+  async assertFormVisible() {
+    assertComponentsVisible([
+      this.createPasswordInput,
+      this.confirmPasswordInput,
+      this.createPasswordLabel,
+      this.confirmPasswordLabel,
+    ]);
+  }
+
+  async assertErrorsHidden() {
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/PasswordPolicyPage.ts:86]: See comment about same thing on name page.
Nothing wrong with this, but since you've made that clever util method for visible, may as well do the same with hidden

File: e2e/pages/StageOne/PasswordPolicyPage.ts [hunk 16]
```diff
@@ -22,4 +58,53 @@ export default class PasswordPolicyPage {
...
+  }
+
+  async assertFillConfirmPasswordField(password2: string) {
+    await this.confirmPasswordInput.click();
+    await this.confirmPasswordInput.fill(password2);
+  }
+
+  // --- ERRORS LOGIC --- //
+  async assertCreatePasswordError(message: string) {
+    await assertErrorUtil(this.createPasswordError, message);
```
Comment (chrismcneill89) [inline e2e/pages/StageOne/PasswordPolicyPage.ts:104]: Se comments in util files about making that util assertion more generic and re-usable

File: e2e/tests/features/journeys/registration/registrationStage1Test.e2e.ts [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+      await namePage.assertErrorsHidden();
+    });
+    await test.step("When I fill only the first name input", async () => {
+      await namePage.firstNameInput.fill("firstName");
+    });
+    await test.step("When I press continue", async () => {
+      await namePage.continueButton.click();
+    });
+    await test.step("I will see an error message correctly rendered", async () => {
+      await assertErrorUtil(
```
Comment (chrismcneill89) [inline e2e/tests/features/journeys/registration/registrationStage1Test.e2e.ts:33]: Would prefer you to use a page method here, which calls the assertErrorUtil method
You could even pass the expected error message in from here if you like.
E.g. 
await namePage.assertLastNameError("Error: Enter your last name")

And on NamePage file:
async assertLastNameError(message: string) {
    await assertComponentWithText(this.lastNameError, message)
  }

File: e2e/tests/features/journeys/registration/registrationStage1Test.e2e.ts [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    });
+    await test.step("When I incorrectly fill out the password fields", async () => {
+      await passwordPolicyPage.createPasswordInput.fill("TestPassword123!");
+      await passwordPolicyPage.confirmPasswordInput.fill("TestWrongPassword");
+    });
+    await test.step("I will click continue", async () => {
+      await passwordPolicyPage.continueButton.click();
+    });
+    await test.step("When I see the correct error message rendered", async () => {
+      await assertErrorUtil(
```
Comment (chrismcneill89) [inline e2e/tests/features/journeys/registration/registrationStage1Test.e2e.ts:129]: Same as the comment above

**Review by chrismcneill89 — COMMENTED** [review]
Hey Thomas very good stuff.
Loving the new utils methods.

Pretty much all the comments are about just using them a little bit better.
Other than that just minor things, quick tidy up and should be good to go.
Great work

File: e2e/hooks/CustomFixtures.ts
Comment (ThomasJackSmolen) [inline e2e/hooks/CustomFixtures.ts:1]: - Addressed comments for stage 1
 - Finished all the POMs for stage 2 
 - Wrote a user journey for stage 2

PROBLEMS 
 - Stage 2 OTP verification doesn't work as a result I can't test the journey

**Review by chrismcneill89 — APPROVED** [review]
Looks good to me, theres still a few pages with numbers in their variables, address that then all clear from me
